[lldb] Handle deref of register and implicit locations

dmpots opened the pull request:
This commit modifies the dwarf expression evaluator in how we handle the deref operation for register and implicit locations on the stack. For a typical memory location a deref operation will read the value from memory. For register and implicit locations the deref operation will read the value from the register or its implicit location. In lldb we eagerly read register and implicit values and push them on the stack so the deref operation for these becomes a "no-op" that leaves the value on the stack and updates the tracked location kind.

The motivation for this change is to handle `DW_OP_deref*` operations on location descriptions as described by the heterogenious debugging [extensions](https://rocm.docs.amd.com/projects/llvm-project/en/latest/LLVM/llvm/html/AMDGPUDwarfExtensionsForHeterogeneousDebugging.html#a-2-5-4-4-4-register-location-description-operations).

Specifically, for register locations it states

> These operations obtain a register location. To fetch the contents of
> a register, it is necessary to use DW_OP_regval_type, use one of the
> DW_OP_breg* register-based addressing operations, or use DW_OP_deref* on
> a register location description.

My understanding is that this is the intended behavior from dwarf5 as well and is not a change in behavior.

clayborg commented on lldb/source/Expression/DWARFExpression.cpp:
```diff
@@ -1106,6 +1124,25 @@ llvm::Expected<Value> DWARFExpression::Evaluate(
+                               /*sign=*/false);
+        stack.back().GetScalar() = scalar;
+        break;
+      }
```
We should reuse the `Evaluate_DW_OP_deref` function call from line 864 and pass in the the deref size as a new argument. So pass in `size` to `Evaluate_DW_OP_deref` for this `DW_OP_deref_size`. For the actual `DW_OP_deref` we can pass in the register size.

dmpots replied:
@clayborg Yeah, Its not nice that we have two nearly identical implementations of deref. There are some differences between them, but they may not be important. For example, DW_OP_deref reads the FileAddress from the process after converting to a load address, but DW_OP_deref_size reads a FileAddress from the target.

Let me create a separate PR to unify the implementations.

dmpots committed:
[lldb] Handle deref of register and implicit locations

This commit modifies how we handle the deref operation for register and
implicit locations on the stack. For a typical memory location a deref
operation will read the value from memory. For register and implicit
locations the deref operation will read the value from the register or
its implicit location. In lldb we eagerly read register and implicit
values and push them on the stack so the deref operation for these
becomes a "no-op" that leaves the value on the stack and updates the
tracked location kind.

The motivation for this change is to handle `DW_OP_deref*` operations on
location descriptions as described by the heterogenious debugging
[extensions](https://rocm.docs.amd.com/projects/llvm-project/en/latest/LLVM/llvm/html/AMDGPUDwarfExtensionsForHeterogeneousDebugging.html#a-2-5-4-4-4-register-location-description-operations).

Specifically, for register locations it states

> These operations obtain a register location. To fetch the contents of
> a register, it is necessary to use DW_OP_regval_type, use one of the
> DW_OP_breg* register-based addressing operations, or use DW_OP_deref*
on
> a register location description.

My understanding is that this is the intended behavior from dwarf5 as
well and is not a change in behavior.

ZaricZoran commented on lldb/unittests/Expression/DWARFExpressionTest.cpp:
```diff
@@ -1217,3 +1217,107 @@ TEST_F(DWARFExpressionMockProcessTestWithAArch, DW_op_deref_no_ptr_fixing) {
+  // converts it to a load address.
+  EXPECT_THAT_EXPECTED(
+      Eval({DW_OP_constu, 0x84, 0xa, DW_OP_stack_value, DW_OP_deref_size, 1}),
+      ExpectLoadAddress(0x4));
```
This doesn't seem right or I am not understanding how the uleb128 output is represented here.

Shouldn't dereferencing of the stack value be the value itself as a load address (aka value of 0x84)? Also what does 0xa byte represents here?

I would expect for the expression to look something more like this:
`DW_OP_constu 0x84 DW_OP_stack_value DW_OP_deref_size 1`
